keps/sig-node/1769-memory-manager/README.md

```diff
@@ -0,0 +1,782 @@
+
+#### Story 2 : Databases
+
+- Databases (e.g., Oracle, PostgreSQL, and MySQL) require relatively extensive amount of memory and hugepages to access massive volumes of data relatively efficiently. In order to reduce the latency caused by cross-NUMA memory access and sharing, all resources (CPU cores, memory, hugepages, and I/O devices) should be aligned to the same single NUMA node. This will serve the stability and performance enhancement significantly.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+#### UX
+
+To increase the UX, the number of new kubelet flags was minimized to a minimum.
+The minimum set of kubelet flags, which is necessary to configure the Memory Manger, is presented in [this](#new-flags-and-configuration-of-the-memory-manager) section.
+
+## Design Details 
+
+### How to enable the guaranteed memory allocation over many NUMA nodes?
+
+The Linux kernel does not support multi-NUMA memory allocation (pinning), to the best of our knowledge, and thus, we assumed that we have to wait for a suitable upgrade in the Linux kernel so as to adopt multi-NUMA memory allocation (pinning) strategy in the *Memory Manager*. However, the community urged us to implement multi-NUMA strategy as something that is really necessary. In fact, we come up with a simple concept and design for multi-NUMA memory allocation (pinning), which allows for guaranteed memory allocation over many NUMA nodes. The main idea is to treat a group of NUMA nodes as an independent unit, and manage those independent units right by the *Memory Manager*.
+
+The groups of NUMA nodes must be disjoint. This is the rule. A disjont set of NUMA groups is exemplified in the figure below. The groups in the figure are disjiont, i.e.: `[0]`, `[1, 2]` and `[3]`. The rule must be obeyed because the groups that overlap basically do not ensure guaranteed memory allocation over multiple NUMA nodes. 
+
+![alt text](./multi_numa.png "Multi-NUMA Partitioning")
+
+For example, the following groups overlap, `[0, 1]`, `[1, 2]` and `[3]`, because they have a common NUMA node indexed by `1`. In other words, if a pair of groups overlap (`[0, 1]` and `[1, 2]`), the memory resources of some group (indexed by `1`) can be prior consumed by another group  (`[0, 1]`), and thus, the other group (`[1, 2]`) eventually starves. 
+
+### The Concept of Node Map and Memory Maps
```

Comment:
In this section, the following text was added to refer to state file. 
`Notice: admins can examine the map though the on-disk state file.`

In section _"Hints Generation for Topology Manager"_, it was added:
`Notice: in the case of failure, admins can examine system logs which provide verbose information about hint generation for pods and containers.`

With respect to _"new instrumentation kubelet will expose"_, there was already a discussion on how to NUMA resources could be exposed via **`/metrics`** endpoint, but it didn't pass the review (`/metrics/cadvisor` is depreciated and there was another technical blocker).   

Slack discussion: https://kubernetes.slack.com/archives/C012XSGFZQE/p1597178915220300?thread_ts=1597170127.219300&cid=C012XSGFZQE

Another thread on Slack: https://kubernetes.slack.com/archives/C012XSGFZQE/p1600954360006200?thread_ts=1600954294.006000&cid=C012XSGFZQE

@derekwaynecarr Have you anything particular and `implementable` in mind to order to offer this _"new instrumentation"_?

Reply:
> in addition, do we want to keep counters to measure how often a pod is rejected due to topology alignment?

As for now, `TopologyAffinityError` can be used as a counter. When [Deployemnt](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) is used and pod gets rejected multiple times, a list of pod rejections is well documented in the system and we can count rejections for every node.

For example:
```
Events:
  Type     Reason                 Age   From                        Message
  ----     ------                 ----  ----                        -------
  Normal   Scheduled              30s   default-scheduler           Successfully assigned default/busybox2 to dell8.localdomain
  Warning  TopologyAffinityError  30s   kubelet, dell8.localdomain  Resources cannot be allocated with Topology locality  
```

We thought about providing more information in `Message` field, so that the cause of rejections can be clear, e.g. due to either memory or CPU misalignment, but it would involve the work on Topology Manager rather than Memory Manager.

@derekwaynecarr Do you have anything particular and `implementable` in mind? (the idea to use `/metrics` wasn't endorsed as described above)

Reply:
Maybe we can defer the instrumentation question based on any experience we uncover in the alpha phase.  I know the pod resources api wants to extend the ability to see the explicit mapping for not just devices but also all other concrete pod resource assignments like cpu.  Instrumentation could be built from that data source if needed as well.  Either way, let's defer this debate to beta phase.  Thanks!